### ISSUE TITLE
AEAD-Encrypt definition

### ISSUE THREAD

**Issue by sayrer** [tlswg/draft-ietf-tls-esni#189]
```
The ClientEncryptedSNI.encrypted_sni value is then computed using the
   usual TLS 1.3 AEAD:

    encrypted_sni = AEAD-Encrypt(key, iv, KeyShareClientHello, ClientESNIInner)
```

However, the TLS 1.3 definition is:

```
AEADEncrypted =
          AEAD-Encrypt(write_key, nonce, additional_data, plaintext)
```

In the ESNI text, it's not clear how `iv` maps to TLS 1.3's `per-record nonce` routine.

**Comment by ekr** [comment]
I'm not following what the problem is here: AEAD-Encrypt is a function that takes four arguments, the second of which is a nonce. In TLS 1.3 record encryption, this nonce value is set as described in S 5.3. In ESNI, it is derived via HKDF in the fashion described in S 5.1.1..

**Comment by sayrer** [comment]
I don't doubt that there is a right answer. It wasn't clear to me whether I should mix in a sequence number to `iv` (maybe zero?).

**Comment by sayrer** [comment]
In other words, the second argument to `AEAD-Encrypt` is in TLS 1.3 is "derived from the sequence number and the client_write_iv or server_write_iv".

As I read, it wasn't clear to me whether I should do nothing to the existing `iv` value, or mix in zero.

**Comment by ekr** [comment]
You don't do either. The function takes a nonce argument. In TLS 1.3, that argument is computed one way, and in ESNI it is computed another way. These are positional functions and the names in scope at the call site are irrelevant.

I think your proposed change would confuse rather than illuminating.

I agree with you that adding test vectors would be useful, and if you want to supply some -- though I suggest you wait until the next version -- in a PR, that would be great.

**Comment by sayrer** [comment]
OK, maybe my proposed solution is not the best, but I don't think the status quo is great. I get the point of "These are positional functions and the names in scope at the call site are irrelevant", but that's actually not quite true in TLS, afaik. The names tend to imply input/output treatment. That's why I got confused.

maybe adjust `ClientInnerNonce` like so:

```
struct {
   uint8 inner_nonce[16];
   PaddedServerNameList realSNI;
} ClientESNIInner;
```

and then:

`nonce = HKDF-Expand-Label(Zx, "esni iv", Hash(ESNIContents), iv_length)`

**Comment by ekr** [comment]
> For example, things called "*_iv" are inputs to nonce in TLS 1.3. 

I don't see how this argues your point. Again, the names are irrelevant.

> In some implementations, the input "Initialization Vector (IV)" [RFC4106] might be a different static type from the nonce in the TLS 1.3 definition.

I don't really see how the implementation types are relevant here.

In any case, given the state of the draft and the reasonable likelihood this will change, I think this decision is well within editor discretion.

**Comment by sayrer** [comment]
> > For example, things called "*_iv" are inputs to nonce in TLS 1.3.
> 
> I don't see how this argues your point. Again, the names are irrelevant.

The TLS spec and the ESNI spec do not have a type system. They do have a serialization spec, and I am not arguing about that. I do think the names could be clearer without inventing a type system.
 
> > In some implementations, the input "Initialization Vector (IV)" [RFC4106] might be a different static type from the nonce in the TLS 1.3 definition.
> 
> I don't really see how the implementation types are relevant here.

Well, I think this perspective leads to reverse engineering C/C++ implementations, rather than precisely specifying the inputs to functions like AEAD-Encrypt.

**Comment by sayrer** [comment]
Maybe it could say something like "The ClientEncryptedSNI.encrypted_sni value is then computed using the usual TLS 1.3 AEAD, but note that the second argument, 'iv', is not combined with a sequence number:"

The reason I found this confusing was that TLS 1.3 doesn't define the "nonce" argument symbolically, but the ESNI draft does define "iv" symbolically. That's fine, but I then had to plumb a new kind of value into the library to allow for this IV, and I wasn't sure if it was right.

**Comment by nharper** [comment]
I agree with @ekr that AEAD-Encrypt is a 4 positional argument function and it shouldn't matter what they're called in the ESNI spec compared to what they're called in RFC 8446. However, from re-reading RFC 8446, I don't see a good definition of the AEAD-Encrypt function. Perhaps it would be better to describe the encryption using the Authenticated Encryption interface described in [Section 2.1 of RFC 5116](https://tools.ietf.org/html/rfc5116#section-2.1)?

**Comment by sayrer** [comment]
"AEAD-Encrypt is a 4 positional argument function and it shouldn't matter what they're called in the ESNI spec compared to what they're called in RFC 8446."

This is sort of true, but I noticed that the specs use the name of the second argument to imply its input treatment: IV vs nonce. In that light, I suppose my suggested text might sound redundant, but I think the current document is too concise.

**Comment by chris-wood** [comment]
Fixed in #194.
